x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/InlineJoin.java

```diff
@@ -142,15 +142,25 @@ public record LogicalPlanTuple(LogicalPlan stubReplacedSubPlan, LogicalPlan orig
      * \_Limit[1000[INTEGER],false]
      *   \_LocalRelation[[x{r}#99],[IntVectorBlock[vector=ConstantIntVector[positions=1, value=1]]]]
      */
-    public static LogicalPlanTuple firstSubPlan(LogicalPlan optimizedPlan) {
+    public static LogicalPlanTuple firstSubPlan(LogicalPlan optimizedPlan, LocalRelation skipped) {
         Holder<LogicalPlanTuple> subPlan = new Holder<>();
         // Collect the first inlinejoin (bottom up in the tree)
         optimizedPlan.forEachUp(InlineJoin.class, ij -> {
             // extract the right side of the plan and replace its source
-            if (subPlan.get() == null && ij.right().anyMatch(p -> p instanceof StubRelation)) {
-                var p = replaceStub(ij.left(), ij.right());
-                p.setOptimized();
-                subPlan.set(new LogicalPlanTuple(p, ij.right()));
+            if (subPlan.get() == null) {
+                if (ij.right().anyMatch(p -> p instanceof StubRelation)) {
+                    var p = replaceStub(ij.left(), ij.right());
+                    p.setOptimized();
+                    subPlan.set(new LogicalPlanTuple(p, ij.right()));
+                } else if (ij.right() instanceof LocalRelation && skipped == null
+                    || ij.right() instanceof LocalRelation == false && ij.right().anyMatch(p -> p instanceof LocalRelation)) {
+                        // In case the plan was optimized further and the StubRelation was replaced with a LocalRelation
+                        // or the right hand side became a LocalRelation alltogether, there is no need to replace the source of the
+                        // right-hand side anymore.
+                        var p = ij.right();
+                        p.setOptimized();
+                        subPlan.set(new LogicalPlanTuple(p, ij.right()));
```

Comment:
Don't know if it'd be worth creating a light static method to do this here and the similar thing above.